crates/red_knot/src/module.rs

```diff
@@ -388,12 +389,70 @@ pub fn file_to_module(db: &dyn SemanticDb, file: FileId) -> QueryResult<Option<M
 //////////////////////////////////////////////////////
 
 /// Changes the module search paths to `search_paths`.
-pub fn set_module_search_paths(db: &mut dyn SemanticDb, search_paths: Vec<ModuleSearchPath>) {
+pub fn set_module_search_paths(db: &mut dyn SemanticDb, search_paths: ResolvedSearchPathOrder) {
     let jar: &mut SemanticJar = db.jar_mut();
 
     jar.module_resolver = ModuleResolver::new(search_paths);
 }
 
+const TYPESHED_STDLIB_DIRECTORY: &str = "stdlib";
+
+/// A resolved module resolution order, implementing PEP 561
+/// (with some small, deliberate differences)
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct ResolvedSearchPathOrder(Vec<ModuleSearchPath>);
+
+impl Deref for ResolvedSearchPathOrder {
+    type Target = [ModuleSearchPath];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// Implementation of the module resolution order from PEP-561.
+///
+/// - `extra_paths` is a list of user-provided paths
+///   that should take first priority in the module resolution.
+///   Examples in other type checkers are mypy's MYPYPATH environment variable,
+///   or pyright's stubPath configuration setting.
+/// - `workspace_root` is the root of the workspace,
+///   used for finding first-party modules
+/// - `site-packages` is the path to the user's `site-packages` directory,
+///   where third-party packages from ``PyPI`` are installed
+/// - `custom_typeshed` is a path to standard-library typeshed stubs.
+///   Currently this has to be a directory that exists on disk.
+///   (TODO: fall back to vendored stubs if no custom directory is provided.)
+impl ResolvedSearchPathOrder {
+    pub fn new(
+        extra_paths: Vec<PathBuf>,
+        workspace_root: PathBuf,
+        site_packages: Option<PathBuf>,
+        custom_typeshed: Option<PathBuf>,
+    ) -> Self {
+        Self(
```

Comment:
I think a slightly nicer API (requires a few more allocations but whathever, we build this once) is to use a builder:

```
ResolvedSearchPathBuilder::workspace(workspace_root)
    .extras(extras)
    .site_packages(site_packages)
    .typeshed(typeshed)
    .build()
```

where `ResolvedSearchPathBuilder` is 

```rust
struct ResolvedSearchPathBuilder {
    extras: Vec<PathBuf>,
    workspace_root: PathBuf,
    typeshed: Option<PathBuf>,
    site_packages: Option<PathBuf>
}
```

But I don't think it realy maters.

Reply:
👍 I considered this, but here I think I prefer to force us to explicitly state whether we're passing in `site-packages` or whether we want to run in "isolated mode", ignoring all `site-packages` packages.

Reply:
I did almost comment on the call to `OrderedSearchPaths::new` above that it's hard to read/understand the call in code review, without IDE support, and without Python keyword arguments. Not sure how much this matters. Another way to address it would be taking in a config struct as argument instead of just positional args.

Reply:
I think we can ignore it for now. Ultimately what we'll do is to read the settings and take the fields from their (or fall back to other defaults).